<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Data-Extraction" data-toc-modified-id="Data-Extraction-1">Data Extraction</a></span><ul class="toc-item"><li><span><a href="#Import-the-necessary-libraries" data-toc-modified-id="Import-the-necessary-libraries-1.1">Import the necessary libraries</a></span></li><li><span><a href="#Import-the-datasets" data-toc-modified-id="Import-the-datasets-1.2">Import the datasets</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#CO2-and-Greenhouse-Gas-Emissions" data-toc-modified-id="CO2-and-Greenhouse-Gas-Emissions-1.2.0.1">CO2 and Greenhouse Gas Emissions</a></span></li><li><span><a href="#Global-Land-Temperature-By-Country" data-toc-modified-id="Global-Land-Temperature-By-Country-1.2.0.2">Global Land Temperature By Country</a></span></li><li><span><a href="#Regarding-Both-Datasets" data-toc-modified-id="Regarding-Both-Datasets-1.2.0.3">Regarding Both Datasets</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Data-Cleaning" data-toc-modified-id="Data-Cleaning-2">Data Cleaning</a></span><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Data-Consideration-1.1:-Identify-the-relevant-columns-in-the-CO2-and-Greenhouse-Gas-Emissions-dataset-and-filter-them-out-into-another-dataframe." data-toc-modified-id="Data-Consideration-1.1:-Identify-the-relevant-columns-in-the-CO2-and-Greenhouse-Gas-Emissions-dataset-and-filter-them-out-into-another-dataframe.-2.0.0.1">Data Consideration 1.1: Identify the relevant columns in the <em>CO2 and Greenhouse Gas Emissions</em> dataset and filter them out into another dataframe.</a></span></li><li><span><a href="#Data-Cleaning-Consideration-2.1:-Change-Dtype-of-dt-column-to-datetime,-and-add-a-new-Year-column" data-toc-modified-id="Data-Cleaning-Consideration-2.1:-Change-Dtype-of-dt-column-to-datetime,-and-add-a-new-Year-column-2.0.0.2">Data Cleaning Consideration 2.1: Change Dtype of <code>dt</code> column to <code>datetime</code>, and add a new <code>Year</code> column</a></span></li><li><span><a href="#Data-Consideration-3.1:-Standardise-the-list-of-countries" data-toc-modified-id="Data-Consideration-3.1:-Standardise-the-list-of-countries-2.0.0.3">Data Consideration 3.1: Standardise the list of countries</a></span></li><li><span><a href="#Data-Cleaning-Consideration-2.2:-Limit-countries-to-those-of-the-Asia-continent" data-toc-modified-id="Data-Cleaning-Consideration-2.2:-Limit-countries-to-those-of-the-Asia-continent-2.0.0.4">Data Cleaning Consideration 2.2: Limit countries to those of the Asia continent</a></span></li><li><span><a href="#Data-Consideration-3.2:-Standardise-the-range-of-years" data-toc-modified-id="Data-Consideration-3.2:-Standardise-the-range-of-years-2.0.0.5">Data Consideration 3.2: Standardise the range of years</a></span></li><li><span><a href="#Data-Cleaning-Consideration-1.2:-Account-for-NA-values-in-gas_asia_clean-dataframe" data-toc-modified-id="Data-Cleaning-Consideration-1.2:-Account-for-NA-values-in-gas_asia_clean-dataframe-2.0.0.6">Data Cleaning Consideration 1.2: Account for NA values in <em>gas_asia_clean</em> dataframe</a></span></li><li><span><a href="#Data-Cleaning-Consideration-2.3:-Account-for-NA-values-in-temp_asia_clean" data-toc-modified-id="Data-Cleaning-Consideration-2.3:-Account-for-NA-values-in-temp_asia_clean-2.0.0.7">Data Cleaning Consideration 2.3: Account for NA values in <em>temp_asia_clean</em></a></span></li></ul></li></ul></li></ul></li></ul></div>

In [141]:
# Installation of side navigation bar
# Type in cmd line:
    # pip install jupyter_contrib_nbextensions && jupyter contrib nbextension install 
# Restart Jupyter Notebook
# Click on Nbextensions
# Check the following boxes: Table of Contents(2) and toc/toc

# Data Extraction

###### This section serves to extract the required data from their respective .csv files and store them into their own dataframes for the purpose of manipulating and analysing the data in a later process.

##### The datasets that will be used are:
* *[CO2 and Greenhouse Gas Emissions](https://nyc3.digitaloceanspaces.com/owid-public/data/co2/owid-co2-data.csv)*
* *[Global Land Temperature By Country](https://www.kaggle.com/datasets/berkeleyearth/climate-change-earth-surface-temperature-data?resource=download&select=GlobalLandTemperaturesByCountry.csv)*

## Import the necessary libraries

In [142]:
# Data Manipulation
import pandas as pd
import numpy as np

# Plotting
import matplotlib.pyplot as plt
import seaborn as sns

<br>

## Import the datasets

#### CO2 and Greenhouse Gas Emissions

In [143]:
df_gas = pd.read_csv('owid-co2-data.csv') 

# Shape of dataset
print(f"Shape of dataset:\n{df_gas.shape}")

# Total number of non-NA values in this dataset
print(f"\nTotal number of non-NA values in dataset:\n{pd.notna(df_gas).sum()}")

# Total number of NA values in this dataset
print(f"\nTotal number of NA values in dataset:\n{df_gas.isnull().sum()}")


# Summary of dataset
print(f"\n\nSummary of dataset:")
print(df_gas.info())


# Descriptive statistical summary of dataset
print(f"\n\nDescriptive statistical summary of dataset:\n{df_gas.describe()}")

Shape of dataset:
(46523, 74)

Total number of non-NA values in dataset:
country                     46523
year                        46523
iso_code                    39862
population                  38574
gdp                         14551
                            ...  
share_global_other_co2       2386
total_ghg                    6149
total_ghg_excluding_lucf     6149
trade_co2                    4259
trade_co2_share              4258
Length: 74, dtype: int64

Total number of NA values in dataset:
country                         0
year                            0
iso_code                     6661
population                   7949
gdp                         31972
                            ...  
share_global_other_co2      44137
total_ghg                   40374
total_ghg_excluding_lucf    40374
trade_co2                   42264
trade_co2_share             42265
Length: 74, dtype: int64


Summary of dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46523 entries, 0 t

> **Data Cleaning Consideration 1.1:** As seen from the result above, the dataset has a total of 74 columns. However for this project, not all of the 74 columns will be utilised. Hence, relevant columns must be filtered out into a separate dataframe. This issue (along with the specific columns that will be utilised) will be addressed in the Data Cleaning portion of this file.

> **Data Cleaning Consideration 1.2:** Because of the number of columns present in this dataset, it is difficult to tell if the columns we intend to utilise contain any NA values, and if so, what proportion of the data in those columns are NA values. Additionally, the years accounted for in this dataset may not be the same as that of the other dataset (ie one dataset may cover years 1975 to 2022 while the other dataset covers years 1990 to 2020). In this case, the range of years that will be used for this project must first be decided upon (**Data Cleaning Consideration 3.2**) before an appropriate course of action can be taken to address the presence of NA values in the relevant columns.



In [144]:
df_gas.head()

country  year iso_code  population  gdp  cement_co2  \
0  Afghanistan  1850      AFG   3752993.0  NaN         NaN   
1  Afghanistan  1851      AFG   3769828.0  NaN         NaN   
2  Afghanistan  1852      AFG   3787706.0  NaN         NaN   
3  Afghanistan  1853      AFG   3806634.0  NaN         NaN   
4  Afghanistan  1854      AFG   3825655.0  NaN         NaN   

   cement_co2_per_capita  co2  co2_growth_abs  co2_growth_prct  ...  \
0                    NaN  NaN             NaN              NaN  ...   
1                    NaN  NaN             NaN              NaN  ...   
2                    NaN  NaN             NaN              NaN  ...   
3                    NaN  NaN             NaN              NaN  ...   
4                    NaN  NaN             NaN              NaN  ...   

   share_global_cumulative_other_co2  share_global_flaring_co2  \
0                                NaN                       NaN   
1                                NaN                       NaN   
2                                NaN                       NaN   
3                                NaN                       NaN   
4                                NaN                       NaN   

   share_global_gas_co2  share_global_luc_co2  share_global_oil_co2  \
0                   NaN                 0.121                   NaN   
1                   NaN                 0.118                   NaN   
2                   NaN                 0.116                   NaN   
3                   NaN                 0.115                   NaN   
4                   NaN                 0.114                   NaN   

   share_global_other_co2  total_ghg  total_ghg_excluding_lucf  trade_co2  \
0                     NaN        NaN                       NaN        NaN   
1                     NaN        NaN                       NaN        NaN   
2                     NaN        NaN                       NaN        NaN   
3                     NaN        NaN                       NaN        NaN   
4                     NaN        NaN                       NaN        NaN   

   trade_co2_share  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  

[5 rows x 74 columns]

In [145]:
df_gas.tail()

country  year iso_code  population           gdp  cement_co2  \
46518  Zimbabwe  2017      ZWE  14751101.0  2.194784e+10       0.469   
46519  Zimbabwe  2018      ZWE  15052191.0  2.271535e+10       0.558   
46520  Zimbabwe  2019      ZWE  15354606.0           NaN       0.570   
46521  Zimbabwe  2020      ZWE  15669663.0           NaN       0.570   
46522  Zimbabwe  2021      ZWE  15993525.0           NaN       0.570   

       cement_co2_per_capita     co2  co2_growth_abs  co2_growth_prct  ...  \
46518                  0.032   9.596          -0.937           -8.899  ...   
46519                  0.037  11.795           2.199           22.920  ...   
46520                  0.037  11.115          -0.681           -5.772  ...   
46521                  0.036  10.608          -0.507           -4.559  ...   
46522                  0.036  11.296           0.688            6.488  ...   

       share_global_cumulative_other_co2  share_global_flaring_co2  \
46518                                NaN                       0.0   
46519                                NaN                       0.0   
46520                                NaN                       0.0   
46521                                NaN                       0.0   
46522                                NaN                       0.0   

       share_global_gas_co2  share_global_luc_co2  share_global_oil_co2  \
46518                   0.0                 0.219                 0.026   
46519                   0.0                 0.211                 0.033   
46520                   0.0                 0.183                 0.030   
46521                   0.0                 0.194                 0.030   
46522                   0.0                 0.190                 0.030   

       share_global_other_co2  total_ghg  total_ghg_excluding_lucf  trade_co2  \
46518                     NaN     115.59                     28.30      0.910   
46519                     NaN     118.22                     30.83      0.771   
46520                     NaN     117.96                     30.53      0.978   
46521                     NaN        NaN                       NaN      1.006   
46522                     NaN        NaN                       NaN        NaN   

       trade_co2_share  
46518            9.486  
46519            6.537  
46520            8.795  
46521            9.481  
46522              NaN  

[5 rows x 74 columns]

<br>

#### Global Land Temperature By Country

In [146]:
df_temp = pd.read_csv('GlobalLandTemperaturesByCountry.csv') 

# Shape of dataset
print(f"Shape of dataset:\n{df_temp.shape}")

# Total number of non-NA values in this dataset
print(f"\nTotal number of non-NA values in dataset:\n{pd.notna(df_temp).sum()}")

# Total number of NA values in this dataset
print(f"\nTotal number of NA values in dataset:\n{df_temp.isnull().sum()}")


# Summary of dataset
print(f"\n\nSummary of dataset:")
print(df_temp.info())


# Descriptive statistical summary of dataset
print(f"\n\nDescriptive statistical summary of dataset:\n{df_temp.describe()}")

Shape of dataset:
(577462, 4)

Total number of non-NA values in dataset:
dt                               577462
AverageTemperature               544811
AverageTemperatureUncertainty    545550
Country                          577462
dtype: int64

Total number of NA values in dataset:
dt                                   0
AverageTemperature               32651
AverageTemperatureUncertainty    31912
Country                              0
dtype: int64


Summary of dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 577462 entries, 0 to 577461
Data columns (total 4 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   dt                             577462 non-null  object 
 1   AverageTemperature             544811 non-null  float64
 2   AverageTemperatureUncertainty  545550 non-null  float64
 3   Country                        577462 non-null  object 
dtypes: float64(2), object(2)
memory usage: 17.6

> **Data Cleaning Consideration 2.1:** As seen from the result above, the `dt` column has a Dtype of `object`. However, the desired Dtype for said column is `datetime`. Additionally, an additional `Year` column should be added to the dataset, on account of the *CO2 and Greenhouse Gas Emissions* data having only annual data. Hence, the Dtype of the `dt` column will be modified to be `datetime` and an additional column `Year` will be added in the Data Cleaning portion.

> **Data Cleaning Consideration 2.2:** Because of the number of countries present in the dataset, the range of years (upon addressing **Data Cleaning Consideration 3.2**) may be limited by the countries involved. In order to allow for a wider range of years, the countries involved in this project will be limited to those in the Asia continent.

> **Data Cleaning Consideration 2.3:** Both the `AverageTemperature` column and `AverageTemperatureUncertainty` column contain NA values. These NA values however only account for approximately 0.5% of the data for each of the two columns. However, there is a possibility that the percentage may change: while both datasets have a column that accounts for the years, the years accounted for may not be the same across both datasets (ie one dataset may cover years 1975 to 2022 while the other dataset covers years 1990 to 2020). In this case, the range of years that will be used for this project must first be decided upon (**Data Cleaning Consideration 3.2**) before an appropriate course of action can be taken to address the presence of NA values in the two columns stated above.

In [147]:
df_temp.head()

dt  AverageTemperature  AverageTemperatureUncertainty Country
0  1743-11-01               4.384                          2.294   Åland
1  1743-12-01                 NaN                            NaN   Åland
2  1744-01-01                 NaN                            NaN   Åland
3  1744-02-01                 NaN                            NaN   Åland
4  1744-03-01                 NaN                            NaN   Åland

In [148]:
df_temp.tail()

dt  AverageTemperature  AverageTemperatureUncertainty  \
577457  2013-05-01              19.059                          1.022   
577458  2013-06-01              17.613                          0.473   
577459  2013-07-01              17.000                          0.453   
577460  2013-08-01              19.759                          0.717   
577461  2013-09-01                 NaN                            NaN   

         Country  
577457  Zimbabwe  
577458  Zimbabwe  
577459  Zimbabwe  
577460  Zimbabwe  
577461  Zimbabwe

<br>

#### Regarding Both Datasets

> **Data Consideration 3.1:** Comparing the first five rows of each dataset to each other, it can be seen that the first country listed in each dataset does not match. Hence, it can be assumed that the list of countries present in one dataset may not be present in another dataset. Hence, for the purpose of this project, countries that are not present in either of the datasets will be filtered out, as there will be insufficient information to make an accurate forcast for those countries.

> **Data Consideration 3.2:** As mentioned in **Data Consideration 2.2** above, the range of years across both datasets may not be the same. Hence, after addressing **Data Consideration 3.1**, the range of years present across both datasets will be compared and a set range of years will be finalised.

<br><br>

# Data Cleaning

##### This section serves to correct structural and/or problematic values in the datasets. In this case, the Data Cleaning process would be to address the Data Cleaning Considerations mentioned in the Data Extraction portion.

#### Data Consideration 1.1: Identify the relevant columns in the *CO2 and Greenhouse Gas Emissions* dataset and filter them out into another dataframe.

The dataset is accompanied by a [codebook](https://github.com/owid/co2-data/blob/master/owid-co2-codebook.csv) that describes each variable present in the dataset, as well as the source of each variable. 
Referencing the codebook, the following columns have been identified to be relevant to this project:
1. `country`
2. `year`
3. `co2_including_luc`
4. `land_use_change_co2`
5. `methane`
6. `nitrous_oxide`

(if time permits: explanation for why the above columns here)

In [149]:
# Copy the required columns into another dataframe
df_gas_filtered = df_gas.loc[:, ["country", "year", "co2_including_luc", "land_use_change_co2", "methane", "nitrous_oxide"]]


# Shape of dataset
print(f"Shape of dataset:\n{df_gas_filtered.shape}")

# Total number of non-NA values in this dataset
print(f"\nTotal number of non-NA values in dataset:\n{pd.notna(df_gas_filtered).sum()}")

# Total number of NA values in this dataset
print(f"\nTotal number of NA values in dataset:\n{df_gas_filtered.isnull().sum()}")


# Summary of dataset
print(f"\n\nSummary of dataset:")
print(df_gas_filtered.info())


# Descriptive statistical summary of dataset
print(f"\n\nDescriptive statistical summary of dataset:\n{df_gas_filtered.describe()}")

Shape of dataset:
(46523, 6)

Total number of non-NA values in dataset:
country                46523
year                   46523
co2_including_luc      24212
land_use_change_co2    39388
methane                 6150
nitrous_oxide           6150
dtype: int64

Total number of NA values in dataset:
country                    0
year                       0
co2_including_luc      22311
land_use_change_co2     7135
methane                40373
nitrous_oxide          40373
dtype: int64


Summary of dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46523 entries, 0 to 46522
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   country              46523 non-null  object 
 1   year                 46523 non-null  int64  
 2   co2_including_luc    24212 non-null  float64
 3   land_use_change_co2  39388 non-null  float64
 4   methane              6150 non-null   float64
 5   nitrous_oxide        6150 non-

<br>

#### Data Cleaning Consideration 2.1: Change Dtype of `dt` column to `datetime`, and add a new `Year` column

In [150]:
# Make a copy of the df_temp dataframe
df_temp_cleaned = df_temp.copy()

# Modify Dtype of dt column
df_temp_cleaned["dt"] = pd.to_datetime(df_temp["dt"], format="%Y-%m-%d")

# Add a new column Year
df_temp_cleaned['Year'] = pd.DatetimeIndex(df_temp_cleaned['dt']).year

# Summary of updated dataset
print(f"\n\nSummary of dataset:")
print(df_temp_cleaned.info())



Summary of dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 577462 entries, 0 to 577461
Data columns (total 5 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   dt                             577462 non-null  datetime64[ns]
 1   AverageTemperature             544811 non-null  float64       
 2   AverageTemperatureUncertainty  545550 non-null  float64       
 3   Country                        577462 non-null  object        
 4   Year                           577462 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(1), object(1)
memory usage: 22.0+ MB
None


In [151]:
df_temp_cleaned.head()

dt  AverageTemperature  AverageTemperatureUncertainty Country  Year
0 1743-11-01               4.384                          2.294   Åland  1743
1 1743-12-01                 NaN                            NaN   Åland  1743
2 1744-01-01                 NaN                            NaN   Åland  1744
3 1744-02-01                 NaN                            NaN   Åland  1744
4 1744-03-01                 NaN                            NaN   Åland  1744

<br>

#### Data Consideration 3.1: Standardise the list of countries

In [152]:
# Get the unique values in df_gas_filtered["country"]
gas_unique_countries = pd.DataFrame(df_gas_filtered["country"].unique(), columns=["country"])
print(f"Countries in df_gas_filtered:\n{gas_unique_countries}")

# Get the unique values in df_temp_cleaned["Country"]
temp_unique_countries = pd.DataFrame(df_temp_cleaned["Country"].unique(), columns=["Country"])
print(f"\nCountries in df_temp_cleaned:\n{temp_unique_countries}")

# Look for countries in gas_unique_countries that don't exist in temp_unique_countries
gas_countries_remove = gas_unique_countries[~gas_unique_countries["country"].isin(temp_unique_countries["Country"])]
print(f"\nCountries in df_gas_filtered that don't exist in df_temp_cleaned:\n{gas_countries_remove}")

# Look for countries in temp_unique_countries that don't exist in gas_unique_countries
temp_countries_remove = temp_unique_countries[~temp_unique_countries["Country"].isin(gas_unique_countries["country"])]
print(f"\nCountries in df_temp_cleaned that don't exist in df_temp_cleaned:\n{temp_countries_remove}")

Countries in df_gas_filtered:
            country
0       Afghanistan
1            Africa
2      Africa (GCP)
3     Aland Islands
4           Albania
..              ...
264  Western Sahara
265           World
266           Yemen
267          Zambia
268        Zimbabwe

[269 rows x 1 columns]

Countries in df_temp_cleaned:
            Country
0             Åland
1       Afghanistan
2            Africa
3           Albania
4           Algeria
..              ...
238  Virgin Islands
239  Western Sahara
240           Yemen
241          Zambia
242        Zimbabwe

[243 rows x 1 columns]

Countries in df_gas_filtered that don't exist in df_temp_cleaned:
                           country
2                     Africa (GCP)
3                    Aland Islands
11             Antigua and Barbuda
16                      Asia (GCP)
17    Asia (excl. China and India)
..                             ...
256   United States Virgin Islands
257  Upper-middle-income countries
260                        Va

In [175]:
# # Remove rows from df_gas_filtered
# gas_countries_unique_clean = df_gas_filtered.loc[gas_countries_remove["country"].isin(df_gas_filtered["country"])]
# print(f"Countries in df_gas_filtered:\n{gas_countries_unique_clean}")

# # Remove rows from df_temp_cleaned
# temp_countries_unique_clean = df_temp_cleaned.loc[temp_countries_remove["Country"].isin(df_temp_cleaned["Country"])]
# print(f"\nCountries in df_temp_cleaned:\n{temp_countries_unique_clean}")



# Remove rows from df_gas_filtered
gas_countries_clean = df_gas_filtered.loc[~df_gas_filtered["country"].isin(gas_countries_remove["country"])]
print(f"Countries in df_gas_filtered:\n{gas_countries_clean['country'].unique()}")

# Remove rows from df_temp_cleaned
temp_countries_clean = df_temp_cleaned.loc[~df_temp_cleaned["Country"].isin(temp_countries_remove["Country"])]
print(f"\nCountries in df_temp_cleaned:\n{temp_countries_clean['Country'].unique()}")


# Both should have the same countries now and therefore the same number of unique countries
print(f"\n\nNumber of unique countries in df_gas_filtered:\n{gas_countries_clean['country'].nunique()}")
print(f"\nNumber of unique countries in df_temp_cleaned:\n{temp_countries_clean['Country'].nunique()}")


Countries in df_gas_filtered:
['Afghanistan' 'Africa' 'Albania' 'Algeria' 'American Samoa' 'Andorra'
 'Angola' 'Anguilla' 'Antarctica' 'Argentina' 'Armenia' 'Aruba' 'Asia'
 'Australia' 'Austria' 'Azerbaijan' 'Bahamas' 'Bahrain' 'Bangladesh'
 'Barbados' 'Belarus' 'Belgium' 'Belize' 'Benin' 'Bhutan' 'Bolivia'
 'Botswana' 'Brazil' 'British Virgin Islands' 'Bulgaria' 'Burkina Faso'
 'Burundi' 'Cambodia' 'Cameroon' 'Canada' 'Cape Verde'
 'Central African Republic' 'Chad' 'Chile' 'China' 'Christmas Island'
 'Colombia' 'Comoros' 'Congo' 'Costa Rica' 'Croatia' 'Cuba' 'Cyprus'
 'Denmark' 'Djibouti' 'Dominica' 'Dominican Republic' 'Ecuador' 'Egypt'
 'El Salvador' 'Equatorial Guinea' 'Eritrea' 'Estonia' 'Ethiopia' 'Europe'
 'Fiji' 'Finland' 'France' 'French Guiana' 'French Polynesia' 'Gabon'
 'Gambia' 'Georgia' 'Germany' 'Ghana' 'Greece' 'Greenland' 'Grenada'
 'Guadeloupe' 'Guatemala' 'Guernsey' 'Guinea' 'Guyana' 'Haiti' 'Honduras'
 'Hong Kong' 'Hungary' 'Iceland' 'India' 'Indonesia' 'Iran' 'Iraq

<br>

#### Data Cleaning Consideration 2.2: Limit countries to those of the Asia continent

*For the purpose of this consideration, an additional dataset [List of Countries by Continent](https://worldpopulationreview.com/country-rankings/list-of-countries-by-continent) will be used to identify the countries that are in Asia*

In [176]:
df_continents = pd.read_csv("country_continents.csv")

# Summary of dataset
print(f"\n\nSummary of dataset:")
print(df_continents.info())


# Descriptive statistical summary of dataset
print(f"\n\nDescriptive statistical summary of dataset:\n{df_continents.describe()}")



Summary of dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 234 entries, 0 to 233
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   place       234 non-null    int64  
 1   pop2023     234 non-null    float64
 2   growthRate  234 non-null    float64
 3   area        234 non-null    float64
 4   country     234 non-null    object 
 5   cca3        234 non-null    object 
 6   cca2        233 non-null    object 
 7   ccn3        234 non-null    int64  
 8   region      234 non-null    object 
 9   subregion   234 non-null    object 
 10  landAreaKm  234 non-null    float64
 11  density     234 non-null    float64
 12  densityMi   234 non-null    float64
 13  Rank        234 non-null    int64  
 14  continent   234 non-null    object 
 15  rank        234 non-null    int64  
dtypes: float64(6), int64(4), object(6)
memory usage: 29.4+ KB
None


Descriptive statistical summary of dataset:
            place    

In [177]:
df_continents.head()

place     pop2023  growthRate       area       country cca3 cca2  ccn3  \
0     12  45606480.0     0.01566  2381741.0       Algeria  DZA   DZ    12   
1     24  36684202.0     0.03077  1246700.0        Angola  AGO   AO    24   
2    204  13712828.0     0.02696   112622.0         Benin  BEN   BJ   204   
3     72   2675352.0     0.01713   582000.0      Botswana  BWA   BW    72   
4    854  23251485.0     0.02548   272967.0  Burkina Faso  BFA   BF   854   

   region                            subregion  landAreaKm   density  \
0  Africa                      Northern Africa   2381741.0   19.1484   
1  Africa    Middle Africa, Sub-Saharan Africa   1246700.0   29.4250   
2  Africa   Western Africa, Sub-Saharan Africa    112760.0  121.6107   
3  Africa  Southern Africa, Sub-Saharan Africa    566730.0    4.7207   
4  Africa   Western Africa, Sub-Saharan Africa    273600.0   84.9835   

   densityMi  Rank continent  rank  
0    49.5943    34    Africa    34  
1    76.2109    42    Africa    42  
2   314.9718    77    Africa    77  
3    12.2266   144    Africa   144  
4   220.1073    59    Africa    59

In [178]:
# Filter out the countries that are in Asia
df_continents_asia = df_continents.loc[df_continents["region"] == "Asia"]
print(f"Total number of countries in Asia: {len(df_continents_asia['country'])}")
df_continents_asia.head()

Total number of countries in Asia: 50


place      pop2023  growthRate      area      country cca3 cca2  ccn3  \
57      4   42239854.0     0.02702  652230.0  Afghanistan  AFG   AF     4   
58     51    2777970.0    -0.00090   29743.0      Armenia  ARM   AM    51   
59     31   10412651.0     0.00527   86600.0   Azerbaijan  AZE   AZ    31   
60     48    1485509.0     0.00902     765.0      Bahrain  BHR   BH    48   
61     50  172954319.0     0.01033  147570.0   Bangladesh  BGD   BD    50   

   region                          subregion  landAreaKm    density  \
57   Asia  South Central Asia, Southern Asia    652230.0    64.7622   
58   Asia                       Western Asia     28470.0    97.5753   
59   Asia                       Western Asia     82646.0   125.9910   
60   Asia      Western Asia, The Middle East       785.0  1892.3682   
61   Asia  Southern Asia, South Central Asia    130170.0  1328.6803   

    densityMi  Rank continent  rank  
57   167.7341    36      Asia    36  
58   252.7201   140      Asia   140  
59   326.3167    90      Asia    90  
60  4901.2335   154      Asia   154  
61  3441.2821     8      Asia     8

In [179]:
# Make a copy of df_temp_cleaned and df_gas_filtered
temp_countries_clean_copy = temp_countries_clean.copy()
gas_countries_clean_copy = gas_countries_clean.copy()


# Filter out the Asian countries from df_temp_cleaned and df_gas_filtered (since they both must have the same countries)
temp_asia_clean = temp_countries_clean_copy.loc[temp_countries_clean_copy["Country"].isin(df_continents_asia["country"])]
gas_asia_clean = gas_countries_clean_copy.loc[gas_countries_clean_copy["country"].isin(df_continents_asia["country"])]

print(f"Total number of countries in temp_asia_clean: {temp_asia_clean['Country'].nunique()}")
print(f"Total number of countries in gas_asia_clean: {gas_asia_clean['country'].nunique()}")

Total number of countries in temp_asia_clean: 44
Total number of countries in gas_asia_clean: 44


In [180]:
temp_asia_clean.head()

dt  AverageTemperature  AverageTemperatureUncertainty  \
3239 1838-04-01              13.008                          2.586   
3240 1838-05-01                 NaN                            NaN   
3241 1838-06-01              23.950                          2.510   
3242 1838-07-01              26.877                          2.883   
3243 1838-08-01              24.938                          2.992   

          Country  Year  
3239  Afghanistan  1838  
3240  Afghanistan  1838  
3241  Afghanistan  1838  
3242  Afghanistan  1838  
3243  Afghanistan  1838

In [181]:
temp_asia_clean.reset_index(drop=True, inplace=True)
temp_asia_clean.head()


gas_asia_clean.reset_index(drop=True, inplace=True)
gas_asia_clean.head()

country  year  co2_including_luc  land_use_change_co2  methane  \
0  Afghanistan  1850                NaN                2.931      NaN   
1  Afghanistan  1851                NaN                2.968      NaN   
2  Afghanistan  1852                NaN                2.968      NaN   
3  Afghanistan  1853                NaN                3.004      NaN   
4  Afghanistan  1854                NaN                3.004      NaN   

   nitrous_oxide  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN

<br>

#### Data Consideration 3.2: Standardise the range of years

In [182]:
# For df_temp_cleaned:
    # For each country, find the range of years available
temp_year_min = 0
temp_year_max = 0

for c in temp_asia_clean["Country"].unique():
    rows_for_c = temp_asia_clean.loc[temp_asia_clean["Country"] == c]
    c_min = rows_for_c['Year'].min()
    c_max = rows_for_c['Year'].max()
    if (c_max == 2013):
        print(c)
    
    if (temp_year_min == 0):
        temp_year_min = c_min
        temp_year_max = c_max
    else:
        # Compare the range of years across all countries to find the min and max that is shared across all countries
        if (c_min > temp_year_min):
            temp_year_min = c_min
        if (c_max < temp_year_max):
            temp_year_max = c_max

print(f"The min and max year for df_temp_cleaned is {temp_year_min} and {temp_year_max} respectively")

Afghanistan
Armenia
Azerbaijan
Bahrain
Bangladesh
Bhutan
Cambodia
China
Georgia
Hong Kong
India
Indonesia
Iran
Iraq
Israel
Japan
Jordan
Kazakhstan
Kuwait
Kyrgyzstan
Laos
Lebanon
Malaysia
Mongolia
Nepal
North Korea
Oman
Pakistan
Philippines
Qatar
Saudi Arabia
Singapore
South Korea
Sri Lanka
Syria
Taiwan
Tajikistan
Thailand
Turkey
Turkmenistan
United Arab Emirates
Uzbekistan
Vietnam
Yemen
The min and max year for df_temp_cleaned is 1876 and 2013 respectively


In [183]:
# rows_for_c = gas_asia_clean["country"].unique()
# # type(rows_for_c[10])


In [184]:
testing = gas_asia_clean.loc[gas_asia_clean["country"] == "Hong Kong"]
# print(testing.head())
testing.reset_index(drop=True, inplace=True)
testing.head()
p_first_index = testing["methane"].first_valid_index()
print(p_first_index)
print(type(None))
# print(testing.iloc[140])
# p_max = testing["year"].iloc[p_first_index]
# print(p_max)

None
<class 'NoneType'>


In [185]:
# testing = gas_asia_clean.loc[gas_asia_clean["country"] == "Hong Kong"]

# type()
# # print(testing.head())
# testing.reset_index(drop=True, inplace=True)
# testing.tail()
# p_first_index = testing["methane"].last_valid_index()
# print(p_first_index)
# print(testing.iloc[140])
# p_max = testing["year"].iloc[p_first_index]
# print(p_max)

In [186]:
# For df_gas_filtered:
    # For each country: 


# gas_dictionary = dict()


# for c in gas_countries_clean["country"].unique():
#     rows_for_c = df_gas_filtered.loc[df_gas_filtered["country"] == c]
    
#     # Nested dictionary
#     predictors = {"co2_including_luc":[], "land_use_change_co2":[], "methane":[], "nitrous_oxide":[]}
    
#     for p in predictors:
#         p_first_index = df_gas_filtered[p].first_valid_index()
#         p_min = df_gas_filtered["year"].iloc[p_first_index]
        
#         p_last_index = df_gas_filtered[p].last_valid_index()
#         p_max = df_gas_filtered["year"].iloc[p_last_index]
        
#         predictors[p].append(p_min)
#         predictors[p].append(p_max)
    
#     gas_dictionary[c] = predictors


missing_predictors = dict() # To identify the countries that do not have data for any of the predictors

# For each predictor
predictors = {"co2_including_luc":[], "land_use_change_co2":[], "methane":[], "nitrous_oxide":[]}
    

for c in gas_asia_clean["country"].unique():
    rows_for_c = gas_asia_clean.loc[gas_asia_clean["country"] == c]
    rows_for_c.reset_index(drop=True, inplace=True)


    # Find the range of years available for each of the predictors:
    for p in predictors:
        p_first_index = rows_for_c[p].first_valid_index()
        p_last_index = rows_for_c[p].last_valid_index()
        
        # Some countries may not have data for some predictors, and hence p_first_index will return None, so:
        if (p_first_index == None or p_last_index == None):
            if c in missing_predictors:
                missing_predictors[c].append(p)
                
            else:
                missing_predictors[c] = [p]
        else:
            p_min = rows_for_c["year"].iloc[p_first_index]
            p_max = rows_for_c["year"].iloc[p_last_index]
        
            if (len(predictors[p]) == 0):
                predictors[p].append(p_min)
                predictors[p].append(p_max)
            else:

                # Compare the range of years across all countries for each of the predictors to find the min and max that is shared across all countries for each predictor
                if (p_min > predictors[p][0]):
                    predictors[p][0] = p_min
                if (p_max < predictors[p][1]):
                    predictors[p][1] = p_max
    
for p in predictors:
    print(f"\nThe min and max year for {p} is {predictors[p][0]} and {predictors[p][1]} respectively")
    

print(f"\nThe following countries do not have data for one or more of the predictors:\n{missing_predictors}")


The min and max year for co2_including_luc is 1970 and 2021 respectively

The min and max year for land_use_change_co2 is 1850 and 2021 respectively

The min and max year for methane is 1990 and 2019 respectively

The min and max year for nitrous_oxide is 1990 and 2019 respectively

The following countries do not have data for one or more of the predictors:
{'Hong Kong': ['methane', 'nitrous_oxide'], 'Taiwan': ['methane', 'nitrous_oxide']}



<br>The results are as follows:

|                      | Min Year    | Max Year   |
| :----:               | :----:      |:----:      | 
| temp_asia_clean      | 1876        |2013        | 
| co2_including_luc    | 1970        |2021        | 
| land_use_change_co2  | 1850        |2021        | 
| methane              | 1990        |2019        | 
| nitrous_oxide        | 1990        |2019        | 

And hence, the range of years would be from 1990 to 2013, for a total of 23 years.

The purpose of standardising the range of years across the predictors is to ensure that the data are all on the same scale and are comparable, hence creating a more accurate time series forecast.

In [187]:
# Standardise the years

# For gas_asia_clean
gas_asia_clean = gas_asia_clean[gas_asia_clean['year'].between(1990, 2013)]

# For temp_asia_clean
temp_asia_clean = temp_asia_clean[temp_asia_clean['Year'].between(1990, 2013)]

In [188]:
gas_asia_clean.head()

country  year  co2_including_luc  land_use_change_co2  methane  \
140  Afghanistan  1990              2.427                0.403     6.74   
141  Afghanistan  1991              2.134                0.220     7.02   
142  Afghanistan  1992              0.163               -1.319     7.13   
143  Afghanistan  1993             -2.250               -3.737     7.21   
144  Afghanistan  1994             -1.807               -3.261     7.47   

     nitrous_oxide  
140           2.84  
141           2.93  
142           2.89  
143           2.93  
144           2.76

In [189]:
gas_asia_clean.tail()

country  year  co2_including_luc  land_use_change_co2  methane  \
7787   Yemen  2009             27.109                2.858    19.44   
7788   Yemen  2010             22.601               -0.550    23.05   
7789   Yemen  2011             20.813                0.623    22.33   
7790   Yemen  2012             20.099               -0.256    19.80   
7791   Yemen  2013             26.061               -1.099    22.36   

      nitrous_oxide  
7787           3.34  
7788           3.45  
7789           3.42  
7790           3.43  
7791           3.62

In [190]:
temp_asia_clean.head()

dt  AverageTemperature  AverageTemperatureUncertainty  \
1821 1990-01-01               1.558                          0.460   
1822 1990-02-01               3.669                          0.244   
1823 1990-03-01               8.402                          0.311   
1824 1990-04-01              14.457                          0.231   
1825 1990-05-01              22.442                          0.316   

          Country  Year  
1821  Afghanistan  1990  
1822  Afghanistan  1990  
1823  Afghanistan  1990  
1824  Afghanistan  1990  
1825  Afghanistan  1990

In [191]:
temp_asia_clean.tail()

dt  AverageTemperature  AverageTemperatureUncertainty Country  \
100963 2013-05-01              31.173                          0.993   Yemen   
100964 2013-06-01              32.325                          2.249   Yemen   
100965 2013-07-01              31.340                          0.804   Yemen   
100966 2013-08-01              30.833                          2.352   Yemen   
100967 2013-09-01                 NaN                            NaN   Yemen   

        Year  
100963  2013  
100964  2013  
100965  2013  
100966  2013  
100967  2013

Additionally, two countries have been identified for the lack of data for one or more predictors. As such, these two countries will be removed from both the *gas_asia_clean* and *temp_asia_clean* dataframes. This is also to ensure that all the data are on the same scale and are comparable (ie no particular predictor will have lesser data being used)

In [192]:
# Removing Hong Kong and Taiwan from the two dataframes

gas_asia_clean = gas_asia_clean.drop(gas_asia_clean[(gas_asia_clean["country"].str.contains("Hong Kong")) | (gas_asia_clean["country"].str.contains("Taiwan"))].index)
temp_asia_clean = temp_asia_clean.drop(temp_asia_clean[(temp_asia_clean["Country"].str.contains("Hong Kong")) | (temp_asia_clean["Country"].str.contains("Taiwan"))].index)

gas_asia_clean.reset_index(drop=True, inplace=True)
temp_asia_clean.reset_index(drop=True, inplace=True)

In [193]:
gas_asia_clean.head()

country  year  co2_including_luc  land_use_change_co2  methane  \
0  Afghanistan  1990              2.427                0.403     6.74   
1  Afghanistan  1991              2.134                0.220     7.02   
2  Afghanistan  1992              0.163               -1.319     7.13   
3  Afghanistan  1993             -2.250               -3.737     7.21   
4  Afghanistan  1994             -1.807               -3.261     7.47   

   nitrous_oxide  
0           2.84  
1           2.93  
2           2.89  
3           2.93  
4           2.76

In [194]:
gas_asia_clean.tail()

country  year  co2_including_luc  land_use_change_co2  methane  \
1003   Yemen  2009             27.109                2.858    19.44   
1004   Yemen  2010             22.601               -0.550    23.05   
1005   Yemen  2011             20.813                0.623    22.33   
1006   Yemen  2012             20.099               -0.256    19.80   
1007   Yemen  2013             26.061               -1.099    22.36   

      nitrous_oxide  
1003           3.34  
1004           3.45  
1005           3.42  
1006           3.43  
1007           3.62

In [195]:
temp_asia_clean.head()

dt  AverageTemperature  AverageTemperatureUncertainty      Country  \
0 1990-01-01               1.558                          0.460  Afghanistan   
1 1990-02-01               3.669                          0.244  Afghanistan   
2 1990-03-01               8.402                          0.311  Afghanistan   
3 1990-04-01              14.457                          0.231  Afghanistan   
4 1990-05-01              22.442                          0.316  Afghanistan   

   Year  
0  1990  
1  1990  
2  1990  
3  1990  
4  1990

In [196]:
temp_asia_clean.tail()

dt  AverageTemperature  AverageTemperatureUncertainty Country  \
11965 2013-05-01              31.173                          0.993   Yemen   
11966 2013-06-01              32.325                          2.249   Yemen   
11967 2013-07-01              31.340                          0.804   Yemen   
11968 2013-08-01              30.833                          2.352   Yemen   
11969 2013-09-01                 NaN                            NaN   Yemen   

       Year  
11965  2013  
11966  2013  
11967  2013  
11968  2013  
11969  2013

<br>

#### Data Cleaning Consideration 1.2: Account for NA values in *gas_asia_clean* dataframe

In [197]:
# Total number of non-NA values in this dataset
print(f"\nTotal number of non-NA values in dataset:\n{pd.notna(gas_asia_clean).sum()}")

# Total number of NA values in this dataset
print(f"\nTotal number of NA values in dataset:\n{gas_asia_clean.isnull().sum()}")


Total number of non-NA values in dataset:
country                1008
year                   1008
co2_including_luc      1008
land_use_change_co2    1008
methane                1008
nitrous_oxide          1008
dtype: int64

Total number of NA values in dataset:
country                0
year                   0
co2_including_luc      0
land_use_change_co2    0
methane                0
nitrous_oxide          0
dtype: int64


As seen above, there are no NA values present in the dataset. Hence, no further cleaning has to be taken.

<br>

#### Data Cleaning Consideration 2.3: Account for NA values in *temp_asia_clean*

In [198]:
# Total number of non-NA values in this dataset
print(f"\nTotal number of non-NA values in dataset:\n{pd.notna(temp_asia_clean).sum()}")


# Total number of NA values in this dataset
print(f"\nTotal number of NA values in dataset:\n{temp_asia_clean.isnull().sum()}")



Total number of non-NA values in dataset:
dt                               11970
AverageTemperature               11928
AverageTemperatureUncertainty    11928
Country                          11970
Year                             11970
dtype: int64

Total number of NA values in dataset:
dt                                0
AverageTemperature               42
AverageTemperatureUncertainty    42
Country                           0
Year                              0
dtype: int64


As seen above, there are 42 rows where there are NA values present in both the `AverageTemperature` and `AverageTemperatureUncertainty` columns. Removing the rows containing the NA values is not viable as because this project intends to use time-series forecasting, if the rows with NA values were to be dropped, there would be missing years for a number of the countries affected, and this in turn will affect the accuracy during training and testing. 
Hence, the alternative solution would be to calculate the mean for each of the columns with NA values (with respect to each country) and subsequently replace the NA values with the mean calculated.

In [199]:
na_values_index = temp_asia_clean[temp_asia_clean["AverageTemperature"].isnull()].index.tolist()

for i in na_values_index:
    # Find country
    c = temp_asia_clean["Country"].iloc[i]
    
    # Get mean of all the AverageTemperature values for that country
    at_mean = temp_asia_clean.loc[:, 'AverageTemperature'].mean()
    
    # Get mean of all the AverageTemperatureUncertainty values for that country
    atu_mean = temp_asia_clean.loc[:, 'AverageTemperatureUncertainty'].mean()
    
    # Replace na value with the mean calculated
    temp_asia_clean.loc[i,'AverageTemperature'] = at_mean
    temp_asia_clean.loc[i,'AverageTemperatureUncertainty'] = atu_mean

In [200]:
# Total number of non-NA values in this dataset
print(f"\nTotal number of non-NA values in dataset:\n{pd.notna(temp_asia_clean).sum()}")

# Total number of NA values in this dataset
print(f"\nTotal number of NA values in dataset:\n{temp_asia_clean.isnull().sum()}")


Total number of non-NA values in dataset:
dt                               11970
AverageTemperature               11970
AverageTemperatureUncertainty    11970
Country                          11970
Year                             11970
dtype: int64

Total number of NA values in dataset:
dt                               0
AverageTemperature               0
AverageTemperatureUncertainty    0
Country                          0
Year                             0
dtype: int64


In [201]:
temp_asia_clean.to_csv('asia_clean.csv')